### ISSUE TITLE
Users no longer able to upload files

### ISSUE THREAD

**Issue by drickerusa** [RocketChat/Rocket.Chat#5940]
Rocket.Chat Version: 0.49.3
Running Instances: 7
DB Replicaset OpLog: yes

This seems to be a recent issue. Users are reporting that uploads are no longer working. 
If I look at the my exception logs I see the following:
![image](https://cloud.githubusercontent.com/assets/21371983/22699768/b15274c4-ed15-11e6-9212-6cd24aafc58c.png)

These errors happen when the user is uploading a file. I also see if when trying to delete a file that I attempted to upload. 
![image](https://cloud.githubusercontent.com/assets/21371983/22699931/1c48f6c2-ed16-11e6-951c-82fe7b50b121.png)

In my two servers both of the /tmp/ufs directories look like this.

![image](https://cloud.githubusercontent.com/assets/21371983/22703329/0d537876-ed21-11e6-850c-59c8cc8bf72a.png)

**Comment by MartinSchoeler** [comment]
@drickerusa  Thanks for reporting this issue. We value your time in helping us to make Rocket.Chat great. We always try to fix confirmed problems as quickly as we can, and it appears that the issue you reported might already be fixed in the latest release. Can you please help us test it against current latest release? Much appreciated. And please remember to backup your data before upgrading, just to be safe. Cheers!

**Comment by drickerusa** [comment]
@MartinSchoeler I'll take an emergency change tonight in order to upgrade to 0.51.0, I'll report back after I've done so.

**Comment by Jamie-Owen** [comment]
@drickerusa We just upgraded to 0.51.0 and I can confirm it's still broken... I recall from a previous issue that we are running a similar setup.

I think I may have found something, as when I looked at the `rocketchat_uploads` collection I noticed a new field: `path` which has the same value as the `url` field minus a `/` like so:

`{ "_id" : "XN4Bdn3z3Yw28dcai",  "path" : "ufs/rocketchat_uploads/XN4Bdn3z3Yw28dcai/IMG.PNG",  "url" : "/ufs/rocketchat_uploads/XN4Bdn3z3Yw28dcai/IMG.PNG" }`

I am able to download the file if I go to this url in the browser, so the upload was in fact successful but it's almost as if the rendering of the upload in the channel is looking for something on the filesystem where it previously used to retrieve it from a fully qualified URL instead.

It's late here, but I'll get around to some more digging tomorrow 👍

**Comment by Jamie-Owen** [comment]
@drickerusa Are you using a reverse proxy? We basically resolved this by setting ip-hash as the load balancing method, where we were previously using least-connected.

Still though, something has changed in recent months, least-connected was working fine for us up until we upgraded last night.

**Comment by drickerusa** [comment]
@Jamie-Owen I upgraded last night to 0.51.0 and was able to upload files now. I do use nginx as the reverse proxy and use the ip_hash setting. I have also tried least-connected. It seems over time I have gone back and forth on these settings (can you use both ?). I have some saml issues that seem to have cropped up after 0.48.1 but I'm not sure if that is the version they showed up after. I'll have to go back and look and create another issue. As far as the upload I'll watch it over the course of the day and reply later to this issue.

**Comment by drickerusa** [comment]
Can I delete the files in /tmp/ufs? I wanted to clean this up.

**Comment by drickerusa** [comment]
RC version 0.51.0
Update: I personally can upload files to the system. but, I just saw an exception error that looks painfully familiar. I have not received any complaints yet. So this may only be one error. Here is the screen.
![image](https://cloud.githubusercontent.com/assets/21371983/22757117/ce49e804-ee06-11e6-9e07-a000183ee537.png)

**Comment by engelgabriel** [comment]
@drickerusa please keep us posted. Is there anyway you can setup a server or a channel on your server that we can use for debugging?

**Comment by drickerusa** [comment]
@engelgabriel Because of security we only allow SSO through SAML to the RC server. Local logins are removed. Its possible that I would have to create a designated user which would have to be treated as a contractor in order for you to use the directory to authenticate to the system.  For right now I'll watch the logs, but as of today I have yet to see another error like that from 2 days ago. It looks as though @Jamie-Owen may have the same issue with the latest version.

**Comment by Jamie-Owen** [comment]
We switched to ip-hash on the reverse proxy and users are no longer complaining about file uploads being broken. As @drickerusa mentioned, there is the odd exception complaining about no such file in `/tmp/ufs` but I haven't had any direct complaints or been able to reproduce it myself.

We're not able to use least-connected load balancing anymore since we upgraded without file upload completely breaking, our version jumped pretty drastically though, we upgraded from 0.37.1 to 0.51.0 so a lot has changed/been introduced. 

The problem seems to stem from the fact that part of the file upload process now relies on something in the filesystem, and using least-connected at the reverse proxy means that any request is likely to not end up at a backend that knows about this file (even more unlikely for us as we're running it in containers too). I would be interested to find out if anyone running a multi-host configuration is experiencing the same issue under similar circumstances.
